### PR TITLE
Merging lots of development

### PR DESCRIPTION
**Pull request by caseyjlaw**

### CODE REVIEW

File: dsacalib/hdf5_io.py [hunk 1]
```diff
@@ -205,3 +205,32 @@ def initialize_hdf5_file(
     t_ds = fhdf.create_dataset("time_seconds", (0, ), maxshape=(None, ),
                                dtype=np.float32, chunks=True, data=None)
     return vis_ds, t_ds
+
```
Comment (danasimard) [inline dsacalib/hdf5_io.py:208]: Moved into sefds.py

File: dsacalib/ms_io.py [hunk 2]
```diff
@@ -449,21 +450,25 @@ def reshape_calibration_data(
...
+    nfreq = vals.shape[-2] if vals is not None else 1
+    npol = vals.shape[-1] if vals is not None else 1
     if np.all(baseline[:ntime*nspw] == baseline[0]):
         if np.all(time[:nspw] == time[0]):
             orig_shape = ['baseline', 'time', 'spw']
             # baseline, time, spw
             time = time.reshape(nbl, ntime, nspw)[0, :, 0]
-            vals = vals.reshape(nbl, ntime, nspw, nfreq, npol)
-            flags = flags.reshape(nbl, ntime, nspw, nfreq, npol)
+            if vals is not None:
```
Comment (danasimard) [inline dsacalib/ms_io.py:523]: Assertion line 453 above means that only one check needs to be done here

File: dsacalib/ms_io.py [hunk 3]
```diff
@@ -472,11 +477,15 @@ def reshape_calibration_data(
             orig_shape = ['baseline', 'spw', 'time']
             assert np.all(spw[:ntime] == spw[0])
             time = time.reshape(nbl, nspw, ntime)[0, 0, :]
-            vals = vals.reshape(nbl, nspw, ntime, nfreq, npol)
-            flags = flags.reshape(nbl, nspw, ntime, nfreq, npol)
+            if vals is not None:
```
Comment (danasimard) [inline dsacalib/ms_io.py:534]: Again, we only need to check that vals is not None here

File: dsacalib/ms_io.py [hunk 4]
```diff
@@ -472,11 +477,15 @@ def reshape_calibration_data(
...
+                vals = vals.reshape(nbl, nspw, ntime, nfreq, npol)
+            if flags is not None:
+                flags = flags.reshape(nbl, nspw, ntime, nfreq, npol)
             if swapaxes:
-                vals = vals.swapaxes(1, 2)
-                flags = flags.swapaxes(1, 2)
+                if vals is not None:
```
Comment (danasimard) [inline dsacalib/ms_io.py:538]: Again, we only need to check that vals is not None here

File: dsacalib/ms_io.py [hunk 5]
```diff
@@ -485,23 +494,31 @@ def reshape_calibration_data(
             # time, baseline, spw
             orig_shape = ['time', 'baseline', 'spw']
             time = time.reshape(ntime, nbl, nspw)[:, 0, 0]
-            vals = vals.reshape(ntime, nbl, nspw, nfreq, npol)
-            flags = flags.reshape(ntime, nbl, nspw, nfreq, npol)
+            if vals is not None:
```
Comment (danasimard) [inline dsacalib/ms_io.py:549]: Again, we only need to check that vals is not None here

File: dsacalib/ms_io.py [hunk 6]
```diff
@@ -485,23 +494,31 @@ def reshape_calibration_data(
...
+                vals = vals.reshape(ntime, nbl, nspw, nfreq, npol)
+            if flags is not None:
+                flags = flags.reshape(ntime, nbl, nspw, nfreq, npol)
             if swapaxes:
-                vals = vals.swapaxes(0, 1)
-                flags = flags.swapaxes(0, 1)
+                if vals is not None:
```
Comment (danasimard) [inline dsacalib/ms_io.py:553]: Again, we only need to check that vals is not None here

File: dsacalib/ms_io.py [hunk 7]
```diff
@@ -801,167 +826,6 @@ def get_antenna_gains(gains, ant1, ant2, refant=0):
                 sign*1.0j*np.angle(gains[idx_phase])))**(-1)
     return antennas, antenna_gains
 
-def write_beamformer_weights(msname, calname, caltime, antennas, outdir,
```
Comment (danasimard) [inline dsacalib/ms_io.py:804]: This function is moved to weights.py

File: dsacalib/ms_io.py [hunk 8]
```diff
@@ -1005,124 +869,10 @@ def get_delays(antennas, msname, calname, applied_delays):
     # delays[flags] = 0
     return newdelays.astype(np.int), flags
 
-def write_beamformer_solutions(
```
Comment (danasimard) [inline dsacalib/ms_io.py:1008]: This is also moved to weights.py

File: dsacalib/ms_io.py [hunk 9]
```diff
@@ -1511,117 +1254,3 @@ def extract_times(UV, ra, dt):
     UV.Nblts = int(idxmax-idxmin)
     assert UV.data_array.shape[0]==UV.Nblts
     UV.Ntimes = UV.Nblts//UV.Nbls
-
```
Comment (danasimard) [inline dsacalib/ms_io.py:1514]: Moved to weights.py

File: dsacalib/routines.py [hunk 10]
```diff
@@ -566,271 +564,6 @@ def calibration_head(obs_params, ant_params, write_to_etcd=False,
         )
     return status
 
-def _gauss_offset(xvals, amp, mean, sigma, offset):
```
Comment (danasimard) [inline dsacalib/routines.py:569]: Moved to sefds.py

File: dsacalib/sefds.py [hunk 11]
```diff
@@ -0,0 +1,542 @@
...
+            idx = int(tidx*(nbls*nspw)+blidx*nspw)
+            try:
+                tmp = tb.CORRECTED_DATA[idx]
+            except IndexError:
+                vis[j, :, :] = np.nan
+                print(f'No data for tidx {tidx}, blidx {blidx}')
+            else:
+                vis[j, :, :] = tmp
+
+    return vis
```
Comment (danasimard) [inline dsacalib/sefds.py]: This is more appropriate in ms_io.py instead of sefds.py

Comment (danasimard) [reply dsacalib/sefds.py]: With that move, it also makes sense to allow the datacolumn to be specified

File: services/calibration_service.py [hunk 12]
```diff
@@ -60,163 +61,6 @@
...
-                assert solns['cal_solutions']['delays'] == \
-                    latest_solns['cal_solutions']['delays']
-                assert solns['cal_solutions']['eastings'] == \
-                    latest_solns['cal_solutions']['eastings']
-            except (AssertionError, FileNotFoundError):
-                beamformer_names.remove(bfname)
-    else:
-        beamformer_names = []
-        latest_solns = None
-    return beamformer_names, latest_solns
```
Comment (danasimard) [inline services/calibration_service.py]: Replaced by look-up of all recent solutions (not just ones at the current dec) in weights.py

File: services/calibration_service.py [hunk 13]
```diff
@@ -60,163 +61,6 @@
...
-                    yamlfile,
-                    Loader=yaml.FullLoader
-                )['cal_solutions']
-            applied_delays = np.array(calibration_params['delays'])*2
-            LOGGER.error(
-                'Error extracting snap delays from uvh5 files. '
-                'Using delays in {0}'.format(current_solns)
-            )
-    return applied_delays
-
```
Comment (danasimard) [inline services/calibration_service.py]: Moved to weights.py

**Review by danasimard — CHANGES_REQUESTED** [review]

File: dsacalib/sefds.py (hunk 11 above)
Comment (caseyjlaw) [reply dsacalib/sefds.py]: Would you like to push a commit to the development branch for this move?

Comment (danasimard) [reply dsacalib/sefds.py]: Yes, I'll do that.

File: dsacalib/ms_io.py (hunk 3 above)
Comment (caseyjlaw) [reply dsacalib/ms_io.py:534]: I think it would be safer for you to you push a commit to this branch with these changes.

**Comment by danasimard** [comment]
@caseyjlaw I think all of these changes are mine, so I'll make them and push to development.  I also noticed that there are some changes I made to the real-time ms creation on ds/dev that aren't captured in development. I'm testing those and will push those to development as well.

File: dsacalib/sefds.py (hunk 11 above)
Comment (danasimard) [reply dsacalib/sefds.py]: I've made these changes.

File: dsacalib/weights.py [hunk 14]
```diff
@@ -0,0 +1,599 @@
...
+    print(f'Using {len(bfnames)} to get gain array of shape {gains.shape}.')
+    return gains
+
+
+def find_good_solutions(bfnames, gains, threshold_ants=60, threshold_angle=10, mask_bothpols=True, plot=False):
+    """ Given names and gain array, calc good set.
+    Returns indices of bfnames argument that are good.
+    """
+
+    refant_ind = ANTENNAS.tolist().index(int(REFANTS[0]))
```
Comment (danasimard) [inline dsacalib/weights.py:117]: Even though this is the antenna that the gains are solved against, I wonder if it is better to use a T antenna here.  The refant is typically an outrigger, but what matters for the search sensitivity is having good T antenna beamformer weights.

**Review by danasimard — APPROVED** [review]

**Pull request merged**